[Bug](topn opt) Fix Two-Phase read when some rowset swept

eldenmoon opened the pull request:
If this is a Two-Phase read query, and we need to delay the release of Rowset by row->update_delayed_expired_timestamp() to expand the lifespan of rowsets. This is necessary to avoid data loss during the second phase reading, where some stale rowsets may be swept and result in missing data.
    
## Proposed changes

Issue Number: close #xxx

<!--Describe your changes.-->

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

eldenmoon commented:
run buildall

yiguolei commented on be/src/runtime/query_context.h:
```diff
@@ -222,6 +228,14 @@ class QueryContext {
+    // and they are automatically released. For example, in two-phase delayed materialization queries,
+    // Rowsets need to be released after the query is finished to avoid premature release during the query process,
+    // which could lead to query exceptions.
+    std::vector<std::shared_ptr<void>> _reference_holder;
```
Do not use a very abstract structure. I think you could use std::shared_ptr<Rowset> and then the maintainer could understand very clear that the query context hold some rowset. Or there maybe some memory leak for example, if some other people add olap scan node to query context. The olap scan node maybe a shared ptr, it owns query context and query context owns olap scannode .

eldenmoon replied:
done

yiguolei commented on be/src/service/internal_service.cpp:
```diff
@@ -1512,11 +1512,18 @@ Status PInternalServiceImpl::_multi_get(const PMultiGetRequest& request,
         if (!rowset) {
-            LOG(INFO) << "no such rowset " << rowset_id;
-            continue;
+            std::optional<RowsetSharedPtr> from_unused_rowset =
```
If it is hold in query context, why not get it from query context?

eldenmoon replied:
done

yiguolei commented on be/src/olap/storage_engine.cpp:
```diff
@@ -1155,10 +1155,15 @@ Status StorageEngine::execute_task(EngineTask* task) {
 
 // check whether any unused rowsets's id equal to rowset_id
-bool StorageEngine::check_rowset_id_in_unused_rowsets(const RowsetId& rowset_id) {
+std::optional<RowsetSharedPtr> StorageEngine::get_rowset_in_unused_rowsets(
```
I think its very bad to get a rowset from gc space. GC is a standalone service. Other logic should not depend on it. The rowset is already hold in query context, I think you should get it from query context.

eldenmoon replied:
done

eldenmoon commented:
run buildall

eldenmoon committed:
[Bug](topn opt) Fix Two-Phase read when some rowset swept

If this is a Two-Phase read query, and we need to delay the release of Rowset by row->update_delayed_expired_timestamp() to expand the lifespan of rowsets. This is necessary to avoid data loss during the second phase reading, where some stale rowsets may be swept and result in missing data.

For rowsets that have been moved to the unused rowsets, they are also needed in second phase reading.

eldenmoon commented:
run buildall

eldenmoon committed:
move delay logic to tablet

eldenmoon commented:
run buildall

yiguolei reviewed (approved):
LGTM

qidaye reviewed (approved):
LGTM
